src/MySqlConnector/MySqlClient/ConnectionPool.cs

```diff
@@ -7,101 +9,191 @@ namespace MySql.Data.MySqlClient
+					{
+						// session is not valid
+						Utility.Dispose(ref session);
+					}
+					else
+					{
+						// session is valid, reset if supported
+						if (m_connectionStringBuilder.ConnectionReset)
+						{
+							await session.ResetConnectionAsync(m_connectionStringBuilder, cancellationToken).ConfigureAwait(false);
+						}
+						// pooled session is ready to be used; return it
+						return session;
+					}
+				}
 
-				if (m_sessions.Count > 0)
-					return m_sessions.Dequeue();
+				session = new MySqlSession(this);
+				await session.ConnectAsync(m_connectionStringBuilder, cancellationToken);
+				return session;
 			}
-			finally
+			catch
 			{
-				m_semaphore.Release();
+				m_session_semaphore.Release();
+				throw;
 			}
-			return null;
 		}
 
-		public bool Return(MySqlSession session)
+		public void Return(MySqlSession session)
 		{
 			try
 			{
-				m_semaphore.Wait();
-
-				// TODO: Dispose oldest connections in the pool first?
-				if (m_sessions.Count >= m_maximumSize)
-					return false;
 				m_sessions.Enqueue(session);
-				return true;
 			}
 			finally
 			{
-				m_semaphore.Release();
+				m_session_semaphore.Release();
 			}
 		}
 
 		public async Task ClearAsync(CancellationToken cancellationToken)
 		{
+			cancellationToken.ThrowIfCancellationRequested();
 			try
 			{
-				await m_semaphore.WaitAsync(cancellationToken).ConfigureAwait(false);
+				// don't let any new connections out of the pool
+				await m_drain_lock.WaitAsync(cancellationToken).ConfigureAwait(false);
 
-				while (m_sessions.Count > 0)
+				// acquire all of the session slots
+				for (var i = 0; i < m_maximumSize; i++)
 				{
-					cancellationToken.ThrowIfCancellationRequested();
+					await m_session_semaphore.WaitAsync(cancellationToken).ConfigureAwait(false);
 
-					var session = m_sessions.Dequeue();
-					await session.DisposeAsync(cancellationToken).ConfigureAwait(false);
+					// clear all of the existing sessions
+					var tasks = new List<Task>();
+					MySqlSession session;
+					while (m_sessions.TryDequeue(out session))
+					{
+						tasks.Add(session.DisposeAsync(cancellationToken));
+					}
+					if (tasks.Count > 0)
+					{
+						await Task.WhenAll(tasks).ConfigureAwait(false);
+					}
 				}
 			}
 			finally
 			{
-				m_semaphore.Release();
+				// release all of the session slots
+				m_session_semaphore.Release(m_maximumSize);
+
+				// re-open the minimum sessions
+				await OpenMinimumSessions();
+
+				// release the master lock
+				m_drain_lock.Release();
 			}
 		}
 
-		public static ConnectionPool GetPool(MySqlConnectionStringBuilder csb)
+		public static async Task<ConnectionPool> GetPoolAsync(MySqlConnectionStringBuilder csb, CancellationToken cancellationToken)
 		{
 			if (!csb.Pooling)
 				return null;
 
-			string key = csb.ConnectionString;
-			lock (s_lock)
+			var key = csb.ConnectionString;
+
+			ConnectionPool pool;
+			// try without the lock
+			if (!s_pools.TryGetValue(key, out pool))
```

Comment:
Ah.  Would you prefer that I change it to a ConcurrentDictionary or just use the Async lock every time?